### ISSUE TITLE
Trailing commas not marked as syntax errors if line breaks present

### ISSUE THREAD

**Issue by bpartridge** [atom/language-json#59]
### Prerequisites

* [ X ] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

#45, which supposedly fixed #11, introduced a regex that, in normal regex-land, should identify trailing commas after the last key-value pair in an object. This is strictly *illegal* syntax per the JSON spec: there is no trailing value-separator in 2.2 in https://www.ietf.org/rfc/rfc4627.txt , regardless of ignored whitespace. Node's implementation of the spec concurs:

```
> JSON.parse('{"a": 1,\n}')
SyntaxError: Unexpected token } in JSON at position 9

> JSON.parse('{"a": 1\n}')
{ a: 1 }
```

However, because [Atom apparently feeds only one line at a time to the regex](https://github.com/atom/first-mate/issues/57), and #45 [handles newlines by placing them inside a lookahead group in a regex](https://github.com/atom/language-json/pull/45/files#diff-04ca07995733e2f89cefb544ab570f91R84) ... if there is a newline between the trailing comma and the end of the object, it will not be matched!

<img width="125" alt="screenshot 2017-05-17 10 52 57" src="https://cloud.githubusercontent.com/assets/20748/26161924/47441156-3af3-11e7-921d-4024f6ed3243.png">
<img width="136" alt="screenshot 2017-05-17 10 53 04" src="https://cloud.githubusercontent.com/assets/20748/26161927/4ade6758-3af3-11e7-915f-77ef2701645a.png">

For anyone handcrafting JSON Schemas or config files (read: everyone these days) that will be read by strict parsers, this is a huge pain, especially since "trivial" config file changes that seem fine in an editor may break things in production, or only after a lengthy test run. More generally, we should strive to be compliant with the spec.

We may need to introduce a new scope in the grammar for the "space between a trailing comma and a closing }", rather than just having it as a lookahead. That way it would match over multiple lines.

### Steps to Reproduce

In Atom console:

`g = atom.grammars.grammarForScopeName('source.json'); JSON.stringify(g.tokenizeLines('{"a":1,\n}'), null, '  ')`

**Expected behavior:** One token should have `invalid.illegal.trailing-dictionary-separator.json` scope.

**Actual behavior:** None of them do.

**Reproduces how often:** All the time.

### Versions

You can get this information from copy and pasting the output of `atom --version` and `apm --version` from the command line. Also, please include the OS and what version of the OS you're running.

### Additional Information

Any additional information, configuration or data that might be necessary to reproduce the issue.

**Comment by bpartridge** [comment]
@not-an-aardvark tagging you for visibility as it was your PR.

**Comment by not-an-aardvark** [comment]
Working on this.

**Comment by not-an-aardvark** [comment]
I ended up getting stuck, so I'll let someone else implement this. I pushed some [tests](https://github.com/not-an-aardvark/language-json/commit/771d437b291f227f0c947e67209d469dc21aa71e) in case those are useful.

**Comment by 50Wliu** [comment]
Yeah I'm not sure if this will be possible given the single line nature of regexes.

**Comment by bpartridge** [comment]
I dug in a bit, and the problem's pretty deep. Despite the simplicity of the JSON spec, I don't think there's any way to detect ALL illegal syntax with an old-style Sublime grammar. The restriction of the Sublime grammar boils down to this:

*In languages where insignificant newlines are permitted, within a scope, it is impossible to condition on how many, or what order in which, patterns are detected.*

For instance, there seems to be fundamentally no way to say, between the two patterns "just a key-value pair" and "a key-value pair preceded by a comma", only one of them is permitted to be the first in a list.

This surfaces in other places as well; for instance, the following doesn't create an error.

<img width="136" alt="screenshot 2017-05-17 16 16 10" src="https://cloud.githubusercontent.com/assets/20748/26178291/626ee572-3b2b-11e7-9775-5172240c5f49.png">

Unfortunate, but I think this is probably a wontfix.

**Comment by 50Wliu** [comment]
Agreed - I'm going to label this as `blocked` for now.

**Comment by pshem** [comment]
I came here to report this after getting burned, only to see this bug marked as blocked. Has anything changed since last year?

**Comment by Arcanemagus** [comment]
It's likely that this is only possible to fix in a better parsing system, such as within [tree-sitter-json](https://github.com/tree-sitter/tree-sitter-json).
